### ISSUE TITLE
False Positive: Handlers that re-interrupt or rethrow are still treated as swallowing interruption.

### ISSUE THREAD

**Issue by Carlson-JLQ** [google/error-prone#5636]
# False Positive: Handlers that re-interrupt or rethrow are still treated as swallowing interruption.

Version
2.48.0
## Checker
- Checker id: `InterruptedExceptionSwallowed`
- Checker description: This checker detects catch blocks that catch a supertype of InterruptedException (like Exception or Throwable) without handling the interruption separately, and methods that can throw InterruptedException but declare a broader throws clause.

## Description of the false positive
`diff/output` marks 3 false positives for `InterruptedExceptionSwallowed`. Across these samples, the checker tends to fire on the outer syntax while missing code-level cues that make the pattern harmless, exempted, or at least much weaker than a real bug. The recurring themes are broad catch handling.

### `NegCase2.java`
Source: `output/sonarqube/InterruptedExceptionCheck/src/main/java/scensct/core/neg/NegCase2.java`

The critical logic sits in a broad `catch` block, so the result depends on whether the checker follows what the handler actually does with interruption or just the exception type it catches. The handler restores the interruption signal, so the interrupt is not actually swallowed. That matches the benchmark's stated intent: Catch block handling Exception and calling Thread.interrupt() should not be flagged as violation.

```java
// Catch block handling Exception and calling Thread.interrupt() should not be flagged as violation.
package scensct.core.neg;

public class NegCase2 {
    public static void main(String[] args) {
        try {
            Thread.sleep(100); // Statement that can throw InterruptedException
        } catch (Exception e) {
            // Scenario 2: Catch Exception, call interrupt, no rethrow, no specific catch for InterruptedException
            Thread.currentThread().interrupt();
        }
    }
}
```

### `NegCase3.java`
Source: `output/sonarqube/InterruptedExceptionCheck/src/main/java/scensct/core/neg/NegCase3.java`

The critical logic sits in a broad `catch` block, so the result depends on whether the checker follows what the handler actually does with interruption or just the exception type it catches. The handler restores the interruption signal, so the interrupt is not actually swallowed. That matches the benchmark's stated intent: Catch block handling Exception without interrupt, but enclosing finally calls interrupt, should not be flagged as violation.

```java
// Catch block handling Exception without interrupt, but enclosing finally calls interrupt, should not be flagged as violation.
package scensct.core.neg;

public class NegCase3 {
    public static void main(String[] args) {
        try {
            try {
                Thread.sleep(100); // Statement that can throw InterruptedException
            } catch (Exception e) {
                // Scenario 3: Catch Exception, no interrupt, no rethrow
                // No action in catch
            }
        } finally {
            // Enclosing finally block calls interrupt
            Thread.currentThread().interrupt();
        }
    }
}
```

### `NegCase4.java`
Source: `output/sonarqube/InterruptedExceptionCheck/src/main/java/scensct/core/neg/NegCase4.java`

The critical logic sits in a broad `catch` block, so the result depends on whether the checker follows what the handler actually does with interruption or just the exception type it catches. The caught exception is rethrown, so the exceptional control flow is preserved rather than silently consumed. That matches the benchmark's stated intent: Catch block handling Exception and rethrowing the caught exception should not be flagged as violation.

```java
// Catch block handling Exception and rethrowing the caught exception should not be flagged as violation.
package scensct.core.neg;

public class NegCase4 {
    public static void main(String[] args) throws Exception {
        try {
            Thread.sleep(100); // Statement that can throw InterruptedException
        } catch (Exception e) {
            // Scenario 4: Catch Exception and rethrow
            throw e;
        }
    }
}
```

## Cause analysis
The rule is documented as follows: This checker detects catch blocks that catch a supertype of InterruptedException (like Exception or Throwable) without handling the interruption separately, and methods that can throw InterruptedException but declare a broader throws clause.

In this set the problem shows up around broad catch handling. The implementation appears to lean too heavily on surface indicators and not enough on what the code actually does.

That is why these reports read like over-approximation. The code often retains a suspicious signature or control-flow skeleton, but once you read the body, the benchmark's claimed exemption is visible in the source itself.

For `InterruptedExceptionSwallowed`, the practical improvement would be to model those source-level exemptions more explicitly instead of reporting every syntactic near-match.

## References
None known.
